src/typetree.jl

```diff
@@ -153,21 +169,11 @@ else
     ismutabletype(T) = isa(T, DataType) && T.mutable
 end
 
-function typetree(@nospecialize(T), ctx, dl, seen=nothing)
+function typetree_inner(@nospecialize(T), ctx, dl, seen::TypeTreeTable)
     if T isa UnionAll || T isa Union || T == Union{} || Base.isabstracttype(T)
         return TypeTree()
     end
 
-    if seen !== nothing && T ∈ seen
-        return TypeTree()
-    end
```

Comment:
I'm a bit concerned about the siblings case here (and not sure if we have a test).

But basically the problem here was if you have
```

struct B
   ...
end

struct A
  lhs::B
  rhs::B
``

The issue was as follows:
we hit this function for A, creating a new map.
We first add B to the seen map via the lhs, computing things.
Next we call this for B again, via the rhs. However if it's the same seen map, we will thing we have already seen B from recursion, when in reality it is because we already computed it from the lhs call.